main_test.go

```diff
@@ -379,72 +377,148 @@ func runTest(t *testing.T, test *mainTest) {
 	}
 }
 
-func expandEscapedFilenames(t *testing.T) error {
-	// Escaped filenames exist because their unescaped forms are invalid on
-	// Windows. We have no choice but to skip these in testing.
-	if runtime.GOOS == "windows" {
-		return nil
-	}
-
-	match, err := filepath.Glob("testdata/*%*")
+func TestEval(t *testing.T) {
+	files, err := filepath.Glob("testdata/*.jsonnet")
 	if err != nil {
-		return err
+		t.Fatal(err)
 	}
-
-	filenameEscapeRE := regexp.MustCompile(`%[0-9A-Fa-f]{2}`)
-
-	for _, input := range match {
-		file := filenameEscapeRE.ReplaceAllStringFunc(input, func(s string) string {
-			code, err := strconv.ParseUint(s[1:], 16, 8)
-			if err != nil {
-				panic(err)
-			}
-			return string([]byte{byte(code)})
-		})
-		if file != input {
-			if err := os.Link(input, file); err != nil {
-				if !os.IsExist(err) {
-					return fmt.Errorf("linking %s -> %s: %v", file, input, err)
-				}
-			}
-			t.Logf("Linked %s -> %s", input, file)
+	tests := make([]mainTest, 0, len(files))
+	for _, input := range files {
+		name := strings.TrimSuffix(input, ".jsonnet")
+		var meta testMetadata
+		if val, exists := metadataForTests[name]; exists {
+			meta = val
 		}
+		tests = append(tests, mainTest{name: name, input: input, golden: name + ".golden", meta: &meta})
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			runTest(t, &test)
+		})
 	}
-
-	return nil
 }
 
-func TestMain(t *testing.T) {
-	flag.Parse()
-
-	if err := expandEscapedFilenames(t); err != nil {
+func withinWorkingDirectory(t *testing.T, dir string) func() error {
+	t.Helper()
+	cwd, err := os.Getwd()
+	if err != nil {
 		t.Fatal(err)
 	}
-
-	var mainTests []mainTest
-	match, err := filepath.Glob("testdata/*.jsonnet")
-	if err != nil {
+	if err := os.Chdir(dir); err != nil {
 		t.Fatal(err)
 	}
+	return func() error {
+		return os.Chdir(cwd)
+	}
+}
 
-	jsonnetExtRE := regexp.MustCompile(`\.jsonnet$`)
+func TestEvalUnusualFilenames(t *testing.T) {
+	// Escaped filenames exist because their unescaped forms are invalid on Windows. We have no
+	// choice but to skip these in testing.
+	if runtime.GOOS == "windows" {
+		return
+	}
 
-	for _, input := range match {
-		// Skip escaped filenames.
-		if strings.ContainsRune(input, '%') {
-			continue
+	// Are we running within "bazel test"?
+	dir := os.Getenv("TEST_TMPDIR")
+	if len(dir) == 0 {
+		var err error
+		if dir, err = ioutil.TempDir("", "jsonnet"); err != nil {
+			t.Fatal(err)
 		}
```

Comment:
- [ ] I realized that the directory created by `ioutil.TempDir` doesn't get deleted automatically at process exit time, so I need to add the following here:
```go
defer os.RemoveAll(dir)
```